File: LintCode_Python/mergeSortedArray.py
# Given two sorted integer arrays A and B, merge B into A as one sorted array.

# Have you met this question in a real interview? Yes
# Example
# A = [1, 2, 3, empty, empty], B = [4, 5]

# After merge, A will be filled as [1, 2, 3, 4, 5]

# Note
# You may assume that A has enough space (size that is greater or equal to m + n)
# to hold additional elements from B. The number of elements initialized in A and
#  B are m and n respectively.

def mergeSortedArray(A, m, B, n):
    if not A and not B:
        return
    i = m - 1
    j = n - 1
    index = m + n - 1
    while i >= 0 and j >= 0:
        if A[i] > B[j]:
            A[index] = A[i]
            i -= 1
        else:
            A[index] = B[j]
            j -= 1
        index -= 1
    while j >= 0:
        A[index] = B[j]
        j -= 1
        index -= 1
    return A

print mergeSortedArray([1, 2, 3, 0, 0], 3, [4, 5], 2)

File: LeetCode_Java/src/MergeSortedArray.java
import java.util.Arrays;

/**
 * Given two sorted integer arrays nums1 and nums2, merge nums2 into nums1 as one sorted array.
 * Note:
 * You may assume that nums1 has enough space (size that is greater or equal to m + n) to hold 
 * additional elements from nums2. The number of elements initialized in nums1 and nums2 are m and 
 * n respectively.
 * 
 * Tags: Array, Two Pointers
 * @author wendi
 *
 */
public class MergeSortedArray {

	/**
	 * Two Pointers: sorted from end to start
	 * @param int[] nums1, int[] nums2
	 * Time: O(m + n)
	 * Space: O(1)
	 */
	public void mergeSortedArray(int[] nums1, int m, int[] nums2, int n) {
		if (nums1 == null || nums1.length == 0 || nums2 == null || nums2.length == 0) return;
		int p1 = m - 1;
		int p2 = n - 1;
		while (p1 >= 0 && p2 >= 0) {
			if (nums1[p1] > nums2[p2]) nums1[p1 + p2 + 1] = nums1[p1--];
			else nums1[p1 + p2 + 1] = nums2[p2--];
		}
		while (p2 >= 0) {
			nums1[p2] = nums2[p2--];
		}
	}
	
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		MergeSortedArray result = new MergeSortedArray();
		int[] nums1 = {3, 5, 0, 0, 0, 0, 0};
		int[] nums2 = {1, 2, 4, 6, 7};
		result.mergeSortedArray(nums1 , 2, nums2 , 5);
		System.out.println(Arrays.toString(nums1));
	}

}
